### PR TITLE
python-ecosys/kpn-senml: Add KPN SenML library.

### PR DESCRIPTION
**Pull request by iabdalkader**

### CODE REVIEW

**Comment by iabdalkader** [comment]
@jimmo If this is merged, and the `require`'d in a `manifest.py`, can I `import kpn_senml` or need to move it up a level ?

**Comment by dpgeorge** [comment]
How does this relate to https://github.com/kpn-iot/senml-python-library and https://github.com/kpn-iot/senml-micropython-library, is it just a copy of the latter?  If so, then I don't think adding it verbatim here is the right way to go.  Instead one can just use the github URL to reference this library.

(All the examples have a boot.py and pymakr.conf which are really not necessary.)

**Comment by dpgeorge** [comment]
Out of interest, what do you need this library for?  Are there other libraries that could be used instead?

> Both libraries seem unmaintained and they don't respond to PRs/issues, and as of now we don't have a way to freeze remote modules in `manifest.py`. If that's implemented soon, we can fork that library instead and freeze it.

I think that's the best solution at this point in time: to fork that repo, apply your patches as new commits, then implement freezing of remote URLs in manifest.py.

**Comment by iabdalkader** [comment]
> Out of interest, what do you need this library for? Are there other libraries that could be used instead?

It's used by cloud clients to parse/create CBOR/JSON. And not aware of another library, but this one works really well.

> I think that's the best solution at this point in time: to fork that repo, apply your patches as new commits, then implement freezing of remote URLs in manifest.py.

I don't mind that. My end goal is just to make devices capable of connecting to the cloud out of the box, and for that I need this library, #508 for time and improved logging in #507 and the cloud client library which won't be in `micropython-lib` anyway, but we don't have remote freezing yet. Any ETA on that @jimmo ?

I think #508 and #507 are generic enough to be in `micropython-lib` though.

**Comment by dpgeorge** [comment]
We need to consider this as two separate components: cbor and senml.  And for both of them we need to consider whether the library is trying to be compatible with an existing library, or if it's a custom MicroPython library.

For cbor: there is an existing `cbor` library on PyPI but it looks abandoned.  Instead there is `cbor2` which had a commit 8 days ago (see https://github.com/agronholm/cbor2).  It looks like the cbor code here (from KPN) is trying to be compatible with `cbor2`.  So if we decide to keep that compatibility then our version should be `python-ecosys/cbor2`.  And I think that's what we should do.

For senml: aside from `kpn_senml`, there doesn't seem to be any other maintained senml library.  If we want to maintain compatibility with `kpn_senml` then our library should be `python-ecosys/kpn_senml`.  Otherwise it should be `micropython/senml` and we are free to change the API.  I would suggest the latter (call it `micropython/senml`) because `kpn_senml` has not been updated in a year, and the PyPI package has not been updated in about 4 years.  Also I think we can remove the special KPN names, we don't need them and they take up space.

**Comment by jimmo** [comment]
Agree to both -- cbor2 and rename to senml, remove all kpn references.

**Comment by iabdalkader** [comment]
> Also I think we can remove the special KPN names, we don't need them and they take up space.
> remove all kpn references

@dpgeorge @jimmo By remove them, you mean rename `KPN_SENML_*` to maybe `SENML_NAME*` (to match `SENML_UNIT_*` ? Because they're used in the examples, if I remove them will have to also remove all the examples that reference them.

**Comment by iabdalkader** [comment]
Hi, what other changes are needed here to move forward with this ?

**Comment by iabdalkader** [comment]
@dpgeorge @jimmo Hi! We need this hosted somewhere so we can link to it and update the imports (`kpn_senml` to `senml` etc..) before publishing our library which needs to be out very soon, what's blocking this PR ?

File: python-ecosys/cbor2/manifest.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+metadata(version="0.0.1")
```
Comment (dpgeorge) [inline python-ecosys/cbor2/manifest.py:1]: I think we can start at version 0.1.0

File: python-ecosys/cbor2/cbor2/decoder.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+break_marker = object()
+
+
+class CBORSimpleValue(object):
+    """
+    Represents a CBOR "simple value".
+    :param int value: the value (0-255)
+    """
+
+    __slots__ = "value"
```
Comment (dpgeorge) [inline python-ecosys/cbor2/cbor2/decoder.py:30]: MicroPython doesn't have slots, so this line can be deleted.

File: python-ecosys/cbor2/cbor2/decoder.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    31: lambda self: break_marker,
+}
+
+
+class CBORDecoder(object):
+    """
+    Deserializes a CBOR encoded byte stream.
+    """
+
+    __slots__ = "fp"
```
Comment (dpgeorge) [inline python-ecosys/cbor2/cbor2/decoder.py:190]: This line can be deleted.

File: python-ecosys/cbor2/cbor2/encoder.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    dict: encode_map,
+}
+
+
+class CBOREncoder(object):
+    """
+    Serializes objects to a byte stream using Concise Binary Object Representation.
+    """
+
+    __slots__ = "fp"  # keep the ram usage down by specifying the allowed fields and nothing more for this class
```
Comment (dpgeorge) [inline python-ecosys/cbor2/cbor2/encoder.py:126]: This line can be deleted.

File: micropython/senml/README.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Introduction
+
+The KPN SenML library helps you create and parse [senml documents](https://tools.ietf.org/html/draft-ietf-core-senml-13)
```
Comment (dpgeorge) [inline micropython/senml/README.md:3]: There shouldn't be any references to KPN in the code or docs.  Eg reword this as "The SenML library ...".

File: micropython/senml/README.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+in both json and cbor format.
+
+# key features
+
+- Object oriented design.
+- built in support for [senml's unit registry](https://tools.ietf.org/html/draft-ietf-core-senml-12#section-12.1)
+- extensible for new data types
+- makes use of (but doesn't restrict to) KPN's predefined list of record names.
```
Comment (dpgeorge) [inline micropython/senml/README.md:11]: I don't think we should include KPN's predefined list of record names.  As per the discussion in the main thread of this PR, this takes up space and is very specificy to KPN.  Users can define and use their own record names.

File: micropython/senml/README.md [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+- direct support to read/write in json and cbor format.
+- automatically adjusts record data with respect to base time, base value & base sum.
+
+Please visit our [docs site](https://kpn-iot.github.io/senml-micropython-library) for more info.
```
Comment (dpgeorge) [inline micropython/senml/README.md:15]: This link is not relevant, so probably just remove this line.

File: micropython/senml/docs/index.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+Welcome to the API documet site for the micro-python version of KPN's SenML library.
```
Comment (dpgeorge) [inline micropython/senml/docs/index.md:1]: Need to reword this to remove KPN reference.

File: micropython/senml/docs/index.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+Check out the [detailed docs](https://kpn-iot.github.io/senml-library/) for an in-depth explanation of all the features in the library.
```
Comment (dpgeorge) [inline micropython/senml/docs/index.md:3]: Need to remove this line.

File: micropython/senml/senml/senml_kpn_names.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+# License: MIT license.
+# Author: Jan Bogaerts
+#
+# sensor names
+
+from senml.senml_unit import enum
+
+
+SenmlNames = enum(
+    KPN_SENML_PRESSURE="pressure",
```
Comment (dpgeorge) [inline micropython/senml/senml/senml_kpn_names.py:17]: As mentioned elsewhere, IMO this file can be removed.  Users can define their own names simply as strings.

File: micropython/senml/manifest.py [hunk 11]
```diff
@@ -0,0 +1,5 @@
+metadata(version="0.0.1")
```
Comment (dpgeorge) [inline micropython/senml/manifest.py:1]: I think this can start at 0.1.0

File: micropython/senml/examples/base.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+#
+# base value example
+
+from senml import *
+import utime as time
+
+
+pack = SenmlPack("device_name")
+temp = SenmlRecord(
+    SenmlNames.KPN_SENML_TEMPERATURE, unit=SenmlUnits.SENML_UNIT_DEGREES_CELSIUS, value=23.5
```
Comment (dpgeorge) [inline micropython/senml/examples/base.py:19]: This `KPN_SENML_TEMPERATURE` constant can be replaced simply with the string `"temperature"`.  Then there's no need for the special KPN names (which don't really have any official meaning, as far as I know).

**Comment by dpgeorge** [comment]
> By remove them, you mean rename KPN_SENML_* to maybe SENML_NAME* (to match SENML_UNIT_* ? Because they're used in the examples, if I remove them will have to also remove all the examples that reference them.

I suggest to just remove all `KPN_SENML_*` names.  The examples can then simply use strings with the appropriate string value.  I don't think there's anything special about the string (the units on the other hand seem to be "official", so can stay as an enum).

File: micropython/senml/README.md (hunk 5 above)
Comment (iabdalkader) [reply micropython/senml/README.md:3]: What about the license header in `*.py` files ?

**Comment by iabdalkader** [comment]
> I suggest to just remove all KPN_SENML_* names. The examples can then simply use strings with the appropriate string value. 

Removed the file and and replaced all references with strings, I also updated the license headers, how does it look now ?

I just have one concern, do we need the `py` files nested like `cbor2/cbor2/decoder.py` or can we move them up one level ?

File: micropython/senml/README.md (hunk 5 above)
Comment (iabdalkader) [reply micropython/senml/README.md]: I updated all the license headers.

**Comment by dpgeorge** [comment]
> I just have one concern, do we need the `py` files nested like `cbor2/cbor2/decoder.py` or can we move them up one level ?

Yes they need to be nested, because it's a package.  So you have it correct.

File: micropython/senml/examples/actuator.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+json_data = (
+    '[{"bn": "device_name", "n":"actuator", "v": 20 }, {"n": "another_actuator", "vs": "a value"}]'
+)
+print(json_data)
+pack.from_json(json_data)
+
+print('[{"bn": "device_name", "n":"temp", "v": 20, "u": "Cel" }]')
+# this represents the cbor json struct: [{-2: "device_name", 0: "temp", 1: "Cel", 2: 20}]
+cbor_data = binascii.unhexlify("81A4216B6465766963655F6E616D65006474656D70016343656C0214")
```
Comment (dpgeorge) [inline micropython/senml/examples/actuator.py:67]: For this use of binascii, it would be simpler to use the now-supported `bytes.fromhex()`.

File: micropython/senml/examples/basic_cbor.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+pack = SenmlPack("device_name")
+
+while True:
+    with SenmlRecord(
+        "test", value=10
+    ) as rec:  # use a with statement to automatically remove the item from the list when it goes out of scope, generate a value for the record
+        pack.add(rec)
+        cbor_val = pack.to_cbor()
+        print(cbor_val)
+        print(ubinascii.hexlify(cbor_val))
```
Comment (dpgeorge) [inline micropython/senml/examples/basic_cbor.py:41]: Could use `cbor_val.hex()` (unless there is a mistmatch of str/bytes bytes here).

File: python-ecosys/cbor2/examples/cbor_test.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+import ubinascii
+
+input = [
+    {"bn": "urn:dev:ow:10e2073a01080063", "u": "Cel", "t": 1.276020076e09, "v": 23.5},
+    {"u": "Cel", "t": 1.276020091e09, "v": 23.6},
+]
+
+data = encoder.dumps(input)
+print(data)
+print(ubinascii.hexlify(data))
```
Comment (dpgeorge) [inline python-ecosys/cbor2/examples/cbor_test.py:38]: Could use `data.hex()`.

File: micropython/senml/LICENSE [hunk 16]
```diff
@@ -0,0 +1,22 @@
+MIT License
```
Comment (dpgeorge) [inline micropython/senml/LICENSE:1]: Now that this license is copied at the top of each file, this LICENSE file itself can be removed (otherwise it's confusing why it's here, no other packages in this repo have one, I don't think).

File: python-ecosys/cbor2/LICENSE [hunk 17]
```diff
@@ -0,0 +1,22 @@
+MIT License
```
Comment (dpgeorge) [inline python-ecosys/cbor2/LICENSE:1]: As above, can remove this file.

File: micropython/senml/examples/actuator.py [hunk 18]
```diff
@@ -0,0 +1,68 @@
+"""
+The MIT License (MIT)
+
+Copyright (c) 2018 KPN (Jan Bogaerts)
+Copyright (c) 2023 MicroPython
```
Comment (dpgeorge) [inline micropython/senml/examples/actuator.py:5]: For this copyright line, please use either your name or Arduino company name (not "MicroPython").

**Comment by dpgeorge** [comment]
For `cbor2/__init__.py`, I think it should contain imports of the decoder and encoder files, so it matches how other packages (eg `senml`) work.

File: micropython/senml/examples/basic_cbor.py (hunk 14 above)
Comment (iabdalkader) [reply micropython/senml/examples/basic_cbor.py:41]: @dpgeorge I have no idea, I never ran those examples, but I changed it as you suggested.

**Comment by iabdalkader** [comment]
> For cbor2/__init__.py, I think it should contain imports of the decoder and encoder files, so it matches how other packages (eg senml) work.

Done.

**Comment by dpgeorge** [comment]
Merged!!
